### PR TITLE
FW fingerprinting: log all fw versions

### PR DESCRIPTION
**Pull request by sshane**
Instead of replacing old FW versions by `(addr, subaddr)` keys, we now also consider the request's brand. We then separately fingerprint on FW returned from each brand's requests.

This removes the ordering problem and allows us to add FW fingerprinting for new brands without worrying it'll affect other cars. Then we can only query subsets of brands once we implement brand filtering.

This also adds logging of which response corresponds to which brand's query. Users will have to check the new `brand` field in the logs to get the appropriate car fw for fingerprint or car port PRs.

### CODE REVIEW

**Comment by sshane** [comment]
Just need to regen routes and it should pass, will do shortly before merging

File: selfdrive/debug/test_fw_query_on_routes.py [hunk 1]
```diff
@@ -97,9 +88,13 @@
             print("not in supported cars")
             break
 
-          fw_versions_dict = build_fw_dict(car_fw)
-          exact_matches = match_fw_to_car_exact(fw_versions_dict)
-          fuzzy_matches = match_fw_to_car_fuzzy(fw_versions_dict)
+          brands = [CP.carName] if len(CP.carName) else SUPPORTED_BRANDS
```
Comment (sshane) [inline selfdrive/debug/test_fw_query_on_routes.py:91]:
```suggestion
          # try fingerprinting with all brands if route didn't fingerprint
          brands = [CP.carName] if len(CP.carName) else SUPPORTED_BRANDS
```

**Comment by sshane** [comment]
Some example routes with the new logging:

Honda Civic 2016 (cfb32f0fb91b173b|2022-07-06--15-47-56):

```python
Ecu: programmedFuelInjection FW: b'37805-5AA-L660\x00\x00' brand: honda
Ecu: transmission FW: b'28101-5CG-A050\x00\x00' brand: honda
Ecu: gateway FW: b'38897-TBA-A020\x00\x00' brand: honda
Ecu: eps FW: b'39990-TBA,A030\x00\x00' brand: honda
Ecu: fwdRadar FW: b'36161-TBC-A030\x00\x00' brand: honda
Ecu: vsa FW: b'57114-TBA-A570\x00\x00' brand: honda
Ecu: srs FW: b'77959-TBA-A030\x00\x00' brand: honda
Ecu: combinationMeter FW: b'78109-TBC-A310\x00\x00' brand: honda
```

Toyota RAV4 2017 (0a48944e30871bb9|2022-07-06--15-49-10):

```python
Ecu: eps FW: b'8965B42083\x00\x00\x00\x00\x00\x00' brand: toyota
Ecu: esp FW: b'F15260R103\x00\x00\x00\x00\x00\x00' brand: toyota
Ecu: dsu FW: b'881514201400\x00\x00\x00\x00' brand: toyota
Ecu: engine FW: b'\x02342Q2100\x00\x00\x00\x00\x00\x00\x00\x0054213000\x00\x00\x00\x00\x00\x00\x00\x00' brand: toyota
Ecu: fwdRadar FW: b'8821F4702100\x00\x00\x00\x00' brand: toyota
Ecu: fwdCamera FW: b'8646F4202100\x00\x00\x00\x00' brand: toyota
```

Kia Optima (018654717bc93d7d|2022-07-06--15-47-07):

```python
Ecu: transmission FW: b'DJF0T16NL1' brand: subaru
Ecu: fwdCamera FW: b'\xf1\x00JFA LKAS AT USA LHD 1.00 1.00 95895-D5001 h32' brand: hyundai
Ecu: transmission FW: b'\xf1\x816U2VA051\x00\x00\xf1\x006U2V0_C2\x00\x006U2VA051\x00\x00DJF0T16NL1\x00\x00\x00\x00' brand: hyundai
Ecu: esp FW: b'\xf1\x00JF ESC \x0b 11 \x18\x030 58920-D5180' brand: hyundai
Ecu: engine FW: b'\x01TJFAJNU06F201H03' brand: toyota
Ecu: engine FW: b'0\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00' brand: mazda
```

File: selfdrive/car/fw_versions.py [hunk 2]
```diff
@@ -284,16 +285,24 @@ def match_fw_to_car_exact(fw_versions_dict):
...
+  # For each brand, attempt to fingerprint using FW returned from its queries
+  for brand in versions.keys():
+    fw_versions_dict = build_fw_dict(fw_versions, filter_brand=brand)
+    matches = match_fw_to_car_exact(fw_versions_dict)
+
+    if allow_fuzzy and len(matches) == 0:
+      matches = match_fw_to_car_fuzzy(fw_versions_dict)
+
+      # Fuzzy match found
+      if len(matches) == 1:
```
Comment (adeebshihadeh) [inline selfdrive/car/fw_versions.py:301]: hmm, this doesn't look quite right. should only be able to fuzzy fingerprint after exhausting all brands for exact matches

File: selfdrive/car/fw_versions.py [hunk 3]
```diff
@@ -284,18 +285,28 @@ def match_fw_to_car_exact(fw_versions_dict):
...
+        matches = match_fw_to_car_exact(fw_versions_dict)
+      else:
+        matches = match_fw_to_car_fuzzy(fw_versions_dict)
 
-  return exact_match, matches
+      if len(matches) == 1:
+        return exact_match, matches
+
+  else:
+    return True, []
```
Comment (sshane) [inline selfdrive/car/fw_versions.py:309]: All the places that use this function (car_helpers, test_fw_fingerprint), they only use the return values if the length of matches is exactly 1. So checking `== 1` above, and returning empty list here changes no behavior.

File: selfdrive/car/fw_versions.py (hunk 2 above)
Comment (sshane) [reply selfdrive/car/fw_versions.py]: Fixed

**Pull request merged**